File: debug_signing.py
#!/usr/bin/env python3
"""
Debug script to output exact EIP-712 signing data for comparison with Rust.
"""

import os
import json
from dotenv import load_dotenv

load_dotenv()

from eth_account import Account
from py_order_utils.builders import OrderBuilder
from py_order_utils.signer import Signer
from py_order_utils.model.order import Order, OrderData
from py_order_utils.model.sides import BUY
from py_order_utils.model.signatures import EOA
from poly_eip712_structs import make_domain
from eth_utils import keccak

PRIVATE_KEY = os.getenv("PRIVATE_KEY")
if not PRIVATE_KEY.startswith("0x"):
    PRIVATE_KEY = "0x" + PRIVATE_KEY

# Constants matching our Rust test case
CHAIN_ID = 137
EXCHANGE = "0xC5d563A36AE78145C45a50134d48A1215220f80a"  # Neg risk exchange
TOKEN_ID = "79003893007240922565581139363959835619617307306268940540301817825959399270354"

# Fixed values for reproducible comparison
SALT = 124398945
MAKER_AMOUNT = 1000000
TAKER_AMOUNT = 2970000

def main():
    print("=" * 70)
    print("EIP-712 Signing Debug")
    print("=" * 70)

    # Create signer
    signer = Signer(PRIVATE_KEY)
    maker_address = signer.address()

    print(f"\nWallet address: {maker_address}")
    print(f"Chain ID: {CHAIN_ID}")
    print(f"Exchange: {EXCHANGE}")

    # Create order with fixed values
    order = Order(
        salt=SALT,
        maker=maker_address,
        signer=maker_address,
        taker="0x0000000000000000000000000000000000000000",
        tokenId=int(TOKEN_ID),
        makerAmount=MAKER_AMOUNT,
        takerAmount=TAKER_AMOUNT,
        expiration=0,
        nonce=0,
        feeRateBps=0,
        side=0,  # BUY
        signatureType=0,  # EOA
    )

    print("\n" + "-" * 70)
    print("Order Data:")
    print("-" * 70)
    print(f"  salt:          {order['salt']}")
    print(f"  maker:         {order['maker']}")
    print(f"  signer:        {order['signer']}")
    print(f"  taker:         {order['taker']}")
    print(f"  tokenId:       {order['tokenId']}")
    print(f"  makerAmount:   {order['makerAmount']}")
    print(f"  takerAmount:   {order['takerAmount']}")
    print(f"  expiration:    {order['expiration']}")
    print(f"  nonce:         {order['nonce']}")
    print(f"  feeRateBps:    {order['feeRateBps']}")
    print(f"  side:          {order['side']}")
    print(f"  signatureType: {order['signatureType']}")

    # Create domain separator (matching Python library)
    domain = make_domain(
        name="Polymarket CTF Exchange",
        version="1",
        chainId=str(CHAIN_ID),
        verifyingContract=EXCHANGE,
    )

    print("\n" + "-" * 70)
    print("Domain:")
    print("-" * 70)
    print(f"  name:              Polymarket CTF Exchange")
    print(f"  version:           1")
    print(f"  chainId:           {CHAIN_ID}")
    print(f"  verifyingContract: {EXCHANGE}")

    # Get signable bytes and hash
    signable = order.signable_bytes(domain=domain)
    struct_hash = keccak(signable)

    print("\n" + "-" * 70)
    print("Hashes:")
    print("-" * 70)
    print(f"  Signable bytes (hex): {signable.hex()}")
    print(f"  Struct hash (hex):    {struct_hash.hex()}")
    print(f"  Struct hash (0x):     0x{struct_hash.hex()}")

    # Sign it
    signature = signer.sign("0x" + struct_hash.hex())
    print(f"\n  Signature: 0x{signature}")

    # Also show the type hash for Order
    order_type_str = "Order(uint256 salt,address maker,address signer,address taker,uint256 tokenId,uint256 makerAmount,uint256 takerAmount,uint256 expiration,uint256 nonce,uint256 feeRateBps,uint8 side,uint8 signatureType)"
    order_type_hash = keccak(order_type_str.encode())
    print(f"\n  Order type hash: 0x{order_type_hash.hex()}")

    # Domain type hash
    domain_type_str = "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
    domain_type_hash = keccak(domain_type_str.encode())
    print(f"  Domain type hash: 0x{domain_type_hash.hex()}")

    print("\n" + "=" * 70)
    print("Use these values to compare with Rust implementation")
    print("=" * 70)

if __name__ == "__main__":
    main()

File: src/bin/debug_signing.rs
//! Debug script to compare EIP-712 signing with Python implementation
//! Run with: cargo run --bin debug_signing

use alloy::dyn_abi::eip712::TypedData;
use alloy::primitives::U256;
use alloy::signers::SignerSync;
use alloy::signers::local::PrivateKeySigner;
use anyhow::Result;
use dotenvy::dotenv;
use std::env;

fn main() -> Result<()> {
    dotenv().ok();
    println!("{}", "=".repeat(70));
    println!("Rust EIP-712 Signing Debug");
    println!("{}", "=".repeat(70));

    // Constants matching Python test case
    let chain_id: u64 = 137;
    let exchange = "0xC5d563A36AE78145C45a50134d48A1215220f80a";
    let maker = "0x83845c1ABC2594Fc28Dfc8A19510EEc03630dFCF";
    let signer = maker;
    let taker = "0x0000000000000000000000000000000000000000";
    let token_id = "79003893007240922565581139363959835619617307306268940540301817825959399270354";

    // Fixed values matching Python
    let salt: u64 = 124398945;
    let maker_amount: u64 = 1000000;
    let taker_amount: u64 = 2970000;
    let expiration: u64 = 0;
    let nonce: u64 = 0;
    let fee_rate_bps: u64 = 0;
    let side: u8 = 0; // BUY
    let signature_type: u8 = 0; // EOA

    // Convert to U256 for display
    let token_id_u256: U256 = token_id.parse().unwrap();

    println!("\nOrder Data:");
    println!("{}", "-".repeat(70));
    println!("  salt:          {}", salt);
    println!("  maker:         {}", maker);
    println!("  signer:        {}", signer);
    println!("  taker:         {}", taker);
    println!("  tokenId:       {}", token_id_u256);
    println!("  makerAmount:   {}", maker_amount);
    println!("  takerAmount:   {}", taker_amount);
    println!("  expiration:    {}", expiration);
    println!("  nonce:         {}", nonce);
    println!("  feeRateBps:    {}", fee_rate_bps);
    println!("  side:          {}", side);
    println!("  signatureType: {}", signature_type);

    println!("\nDomain:");
    println!("{}", "-".repeat(70));
    println!("  name:              Polymarket CTF Exchange");
    println!("  version:           1");
    println!("  chainId:           {}", chain_id);
    println!("  verifyingContract: {}", exchange);

    // Build the EIP-712 TypedData JSON - matching our lib.rs implementation
    let json_str = format!(
        concat!(
            r#"{{"types":{{"EIP712Domain":["#,
            r#"{{"name":"name","type":"string"}},"#,
            r#"{{"name":"version","type":"string"}},"#,
            r#"{{"name":"chainId","type":"uint256"}},"#,
            r#"{{"name":"verifyingContract","type":"address"}}"#,
            r#"],"Order":["#,
            r#"{{"name":"salt","type":"uint256"}},"#,
            r#"{{"name":"maker","type":"address"}},"#,
            r#"{{"name":"signer","type":"address"}},"#,
            r#"{{"name":"taker","type":"address"}},"#,
            r#"{{"name":"tokenId","type":"uint256"}},"#,
            r#"{{"name":"makerAmount","type":"uint256"}},"#,
            r#"{{"name":"takerAmount","type":"uint256"}},"#,
            r#"{{"name":"expiration","type":"uint256"}},"#,
            r#"{{"name":"nonce","type":"uint256"}},"#,
            r#"{{"name":"feeRateBps","type":"uint256"}},"#,
            r#"{{"name":"side","type":"uint8"}},"#,
            r#"{{"name":"signatureType","type":"uint8"}}"#,
            r#"]}},"primaryType":"Order","#,
            r#""domain":{{"name":"Polymarket CTF Exchange","version":"1","chainId":{},"verifyingContract":"{}"}},"#,
            r#""message":{{"salt":"{}","maker":"{}","signer":"{}","taker":"{}","tokenId":"{}","makerAmount":"{}","takerAmount":"{}","expiration":"{}","nonce":"{}","feeRateBps":"0","side":{},"signatureType":{}}}}}"#
        ),
        chain_id,
        exchange,
        salt,
        maker,
        signer,
        taker,
        token_id_u256,
        maker_amount,
        taker_amount,
        expiration,
        nonce,
        side,
        signature_type
    );

    println!("\nTypedData JSON:");
    println!("{}", "-".repeat(70));
    // Pretty print JSON
    let json_val: serde_json::Value = serde_json::from_str(&json_str)?;
    println!("{}", serde_json::to_string_pretty(&json_val)?);

    // Parse and compute hash
    let typed: TypedData = serde_json::from_str(&json_str)?;
    let hash = typed.eip712_signing_hash()?;

    println!("\nHashes:");
    println!("{}", "-".repeat(70));
    println!("  Struct hash (Rust): 0x{}", hex::encode(hash.as_slice()));

    // Python's expected hash for comparison
    let python_hash = "fdfe0dfe96ac85387202046e3dc500babf224fd9d471c1e3a93207aee10115d7";
    println!("  Struct hash (Python): 0x{}", python_hash);

    if hex::encode(hash.as_slice()) == python_hash {
        println!("\n  ✅ HASHES MATCH!");
    } else {
        println!("\n  ❌ HASHES DO NOT MATCH!");
    }

    // Now sign with the private key
    let private_key = env::var("PRIVATE_KEY").expect("PRIVATE_KEY not set");
    let key_clean = private_key.trim().trim_start_matches("0x");
    let key_with_prefix = format!("0x{}", key_clean);
    let wallet: PrivateKeySigner = key_with_prefix.parse()?;

    println!("\n  Wallet address: {}", wallet.address());

    let sig = wallet.sign_hash_sync(&hash)?;
    let sig_hex = format!("0x{}", sig);

    println!("  Signature (Rust):   {}", sig_hex);

    // Python's expected signature
    let python_sig = "0xa0f60ecabcdf2ed8e4cb3b6049cfc641d7be9c7241b195c5e20f149f27fec6823197e5fefbfb4b2ba71c50a52826f56ef90b2694a151b76958664764655fb3e01c";
    println!("  Signature (Python): {}", python_sig);

    if sig_hex.to_lowercase() == python_sig.to_lowercase() {
        println!("\n  ✅ SIGNATURES MATCH!");
    } else {
        println!("\n  ❌ SIGNATURES DO NOT MATCH!");
    }

    println!("\n{}", "=".repeat(70));

    Ok(())
}
